cases/src/org/commcare/cases/util/CasePurgeFilter.java

```diff
@@ -132,61 +126,91 @@ private void setIdsToRemoveNew(IStorageUtilityIndexed<Case> caseStorage, Vector<
+            }
+        }
+        walk(g,false, STATUS_AVAILABLE, STATUS_AVAILABLE, CaseIndex.RELATIONSHIP_EXTENSION);
+    }
+
+    private void markLive(DAG<String, int[], String> g) {
+        for(Enumeration<String> e = g.getIndices(); e.hasMoreElements();) {
+            String index = e.nextElement();
+            int[] node = g.getNode(index);
+            if(is(node, STATUS_OWNED | STATUS_RELEVANT | STATUS_AVAILABLE)) {
+                node[0] |= STATUS_ALIVE;
+            }
+        }
+
+        walk(g, true, STATUS_ALIVE, STATUS_ALIVE);
+        walk(g, false, STATUS_ALIVE, STATUS_ALIVE, CaseIndex.RELATIONSHIP_EXTENSION);
+    }
+
+    private boolean hasOutgoingExtension(DAG<String, int[], String> g, String index) {
+        for(Edge<String, String> edge : g.getChildren(index)) {
+            if(edge.e.equals(CaseIndex.RELATIONSHIP_EXTENSION)) {
+                return true;
             }
         }
+        return false;
+    }
 
-        //Pass 2: Start at Sinks (nodes which only receive edges) and walk down the
-        //graphs, marking necessary nodes as alive
-        toProcess = g.getSinks();
+    private void walk(DAG<String, int[], String> g, boolean direction, int mask, int mark) {
+        walk(g, direction, mask, mark, null);
+    }
+
+    private void walk(DAG<String, int[], String> g, boolean direction, int mask, int mark, String relationship) {
+        Stack<String> toProcess = direction ? g.getSources() : g.getSinks();
         while (!toProcess.isEmpty()) {
             //current node
             String index = toProcess.pop();
-            String[] node = g.getNode(index);
-
-            //Walk all indexing nodes.
-            for (Edge<String, String> edge : g.getParents(index)) {
-                if (g.getNode(edge.i)[0].equals(STATUS_ABANDONED) &&
-                        edge.e.equals(CaseIndex.RELATIONSHIP_EXTENSION) &&
-                        node[0].equals(STATUS_LIVE)) {
+            int[] node = g.getNode(index);
 
-                    g.getNode(edge.i)[0] = STATUS_LIVE;
+            for (Edge<String, String> edge : (direction ? g.getChildren(index) : g.getParents(index))) {
+                if (is(node, mask) && (relationship == null || edge.e.equals(relationship))) {
+                    g.getNode(edge.i)[0] |= mark;
                 }
-
-                //add this to the process stack
-                //TODO: We're going to walk stuff redundantly this
-                //way. Can we optimize?
                 toProcess.addElement(edge.i);
             }
         }
+    }
 
-        //Ok, so now just go through all nodes and signal that we need to remove anything
-        //that isn't live!
-        for (Enumeration iterator = g.getNodes(); iterator.hasMoreElements(); ) {
-            String[] node = (String[])iterator.nextElement();
-            if (!node[0].equals(STATUS_LIVE)) {
-                idsToRemove.addElement(Integer.valueOf(node[1]));
+    private boolean is(int[] node, int flag) {
```

Comment:
consider passing in the case status instead of the int array whose first element is the case's status